### PR TITLE
Swift Code Update

### PR DESCRIPTION
**Pull request by s3482247**
In this PR:
- Added loading indicators for location subviews / event view
- Event annotations have been updated to ensure they are anchored correctly
- Map resource requirements have been reduced by using reusable images
- Logic has been added to control when the map refreshes, reducing lag
- Implemented custom map panning so events/suburb labels are panned to when selected
- Adhered to Mapbox data sharing and Dark Skys attribution requirements
- Added support for managing and persisting user settings
- Added support for filtering events based on a user-configurable category whitelist
- Added support for fetching cached events in the location events list
- Added support for sorting and grouping the location events list
- Event date & times are now shown in the location events list
- Updated bundle identifier to com.peterweller.eventguardian for TestFlight deployment
- Added documentation for api, model, view model, view controller classes & extensions
- Code refactoring to make the coding style more consistent

How i feel after making such a large pull request:
Digging Grave

### CODE REVIEW

File: CPT331.iOS/CPT331.iOS/CrimeManager.swift [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+import Alamofire
+import SwiftyJSON
+
+/// Handles retrieval and parsing of event data from the EventGuardian JSON API
+class CrimeManager: JSONAPI {
+    static let sharedInstance = CrimeManager()
+    
+    // TODO: move domain to info.plist
+    /// EventGuardian API endpoint for crime data (read-only)
+    private let ENDPOINT = "http://ec2-52-32-105-85.us-west-2.compute.amazonaws.com/api/Crime/CrimesByCoordinate"
```
Comment (6nop) [inline CPT331.iOS/CPT331.iOS/CrimeManager.swift]: I think we should move at least up to the host name into a configuration file.

eg ```ENDPOINT = hypotheticalConfigurationClass.EventGuardianApi + "/api/Crime/CrimesByCoordindate"```

This is for two reasons:
 1. It allows us to easily change the production environment without having to update and recompile code. The fact is that the hostname ```ec2-52-32-105-85.us-west-2.compute.amazonaws.com``` is it self a temporary site.
 2. It allows the implementation of staged deployments. 
   * You could deploy the API and App to a staging environment by transforming the configs for staging.
   * UAT then takes place on staging.
   * Then the API and App are promoted to production, driven by another config transform.
   * The key point is there were no code changes between staging/production. Furthermore, the API / App wont even require recompiling.

I think configuration should be easily achievable within the Swift framework.

Comment (s3364859) [inline CPT331.iOS/CPT331.iOS/CrimeManager.swift]: What @6nop said.

File: CPT331.iOS/CPT331.iOS/Dictionary.swift [hunk 2]
```diff
@@ -8,8 +8,11 @@
...
 import Foundation
 
-// Dictionary union
-// Source: http://stackoverflow.com/a/24052094
+/**
+    Merges the dictionary on the right side the operator into the dictionary on the left side.
+    
+    [Third party code](http://stackoverflow.com/a/24052094) by Stack Overflow user: shucao.
+*/
 func += <K, V> (inout left: [K:V], right: [K:V]) {
```
Comment (s3364859) [inline CPT331.iOS/CPT331.iOS/Dictionary.swift]: That is some nasty looking syntax.

File: CPT331.iOS/CPT331.iOS/EventManager.swift [hunk 3]
```diff
@@ -8,34 +8,38 @@
...
 import Alamofire
 import SwiftyJSON
 
+/// Handles retrieval, parsing and caching of event data from the EventGuardian JSON API
 class EventManager: JSONAPI {
     static let sharedInstance = EventManager()
     
     // TODO: move domain to info.plist
+    /// EventGuardian API endpoint for event data (read-only)
     private let ENDPOINT = "http://ec2-52-32-105-85.us-west-2.compute.amazonaws.com/api/Event/EventsByCoordinate"
```
Comment (s3364859) [inline CPT331.iOS/CPT331.iOS/EventManager.swift]: See above comment regarding moving this to some kind of configuration file.

File: CPT331.iOS/CPT331.iOS/LocationManager.swift [hunk 4]
```diff
@@ -7,23 +7,31 @@
...
 class LocationManager {
     static let sharedInstance = LocationManager()
     
-    // Stores a referance to the most recent search task
-    //      To be cancelled if a new task starts
+    /// Stores a referance to the most recent search task. To be cancelled if a new task starts.
     private var lastSearchTask: NSURLSessionDataTask?
     
-    // Prevent external initialization
+    /// Singleton initializer
```
Comment (s3364859) [inline CPT331.iOS/CPT331.iOS/LocationManager.swift]: Probably don't need this level of detail in the comments. Btw, what's the difference between double and triple slashes? Seems as though there's a sprinkling of the two throughout?

File: CPT331.iOS/CPT331.iOS/LocationManager.swift [hunk 5]
```diff
@@ -36,14 +44,19 @@ class LocationManager {
             return
         }
         
+        // Input validation, string must only contain: a-z A-Z 0-9 - . , \s
+        guard query.matchPattern("^[a-zA-Z0-9\\-\\.,\\s']*$") else {
```
Comment (s3364859) [inline CPT331.iOS/CPT331.iOS/LocationManager.swift]: Does regex perform better on iOS than it does everywhere else?

File: CPT331.iOS/CPT331.iOS/NSDate.swift [hunk 6]
```diff
@@ -10,26 +10,119 @@ import Foundation
...
+    }
+}
+
+
+/**
+    Returns true if the LHS date occurs after the RHS date
+ 
+    [Third party code](http://stackoverflow.com/a/27347777) by Stack Overflow user: Andrew.
+ */
+func > (lhs: NSDate, rhs: NSDate) -> Bool {
```
Comment (s3364859) [inline CPT331.iOS/CPT331.iOS/NSDate.swift]: I'm not sure you have to worry about attributing this mate: anyone can overload operators. It's not as if this person supplied a whole framework that you are using.

File: CPT331.iOS/CPT331.iOS/NSDate.swift [hunk 7]
```diff
@@ -10,26 +10,119 @@ import Foundation
...
+    return lhs.timeIntervalSinceReferenceDate > rhs.timeIntervalSinceReferenceDate
+}
+
+
+/**
+    Returns true if the LHS date occurs before the RHS date
+ 
+    [Third party code](http://stackoverflow.com/a/27347777) by Stack Overflow user: Andrew.
+ */
+func < (lhs: NSDate, rhs: NSDate) -> Bool {
```
Comment (s3364859) [inline CPT331.iOS/CPT331.iOS/NSDate.swift]: Same here. I think it'd be like us continuing to attribute any operator overloads to the first person who wrote them:

//    Written by Mr. Man originally in 1965:
bool < (int a, int b)

File: CPT331.iOS/CPT331.iOS/String.swift [hunk 8]
```diff
@@ -9,9 +9,53 @@
...
+    
+    
+    /**
+        Encodes the string as Base64
+     
+        [Third party code](http://stackoverflow.com/a/32512122) by Stack Overflow user: PiXeL16, 2015.
+     
+        - Returns: Base64 encoded string
+    */
     func toBase64() -> String {
```
Comment (s3364859) [inline CPT331.iOS/CPT331.iOS/String.swift]: See previous comments.

File: CPT331.iOS/CPT331.iOS/String.swift [hunk 9]
```diff
@@ -9,9 +9,53 @@
...
+        
+        [Third party code](http://stackoverflow.com/a/34522960) by Stack Overflow user: solos, 2016.
+     
+        - Parameters:
+            - pattern: the regular expression pattern to test
+     
+        - Returns: TRUE if the pattern matches, otherwise FALSE
+    */
+    func matchPattern(regexPattern:String)->Bool {
+
```
Comment (s3364859) [inline CPT331.iOS/CPT331.iOS/String.swift]: See previous comments.

File: CPT331.iOS/CPT331.iOS/WeatherDataCurrent.swift [hunk 10]
```diff
@@ -7,23 +7,22 @@
...
 
 import Foundation
-import SwiftyJSON
 
-class WeatherDataCurrent: WeatherData {
+/// Stores data for weather data at a specific point in time. Extends WeatherData to add support for storing a temperature value and precipitation data.
+class WeatherDataCurrent:WeatherData {
     
-    // The air temperature in degrees Celcius
+    /// The current air temperature in degrees Celcius
```
Comment (s3364859) [inline CPT331.iOS/CPT331.iOS/WeatherDataCurrent.swift]: Which is the only correct way to measure it.

**Comment by s3364859** [comment]
I'm happy with this if we can get that endpoint location moved into configuration rather than having it baked into the code somewhere.

File: CPT331.iOS/CPT331.iOS/Config.plist [hunk 11]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
+	<key>API</key>
+	<dict>
+		<key>DarkSky</key>
+		<string>https://api.darksky.net</string>
+		<key>EventGuardian</key>
```
Comment (6nop) [inline CPT331.iOS/CPT331.iOS/Config.plist]: Thumbs up

File: CPT331.iOS/CPT331.iOS/WeatherManager.swift [hunk 12]
```diff
@@ -17,7 +17,7 @@ class WeatherManager: JSONAPI {
     
     // TODO: move domain and API key to info.plist
     /// DarkSky API endpoint for weather data (read-only)
-    private let ENDPOINT = "https://api.darksky.net/forecast/5136e36ec5498cfd8c583e3f580d3060"
+    private let ENDPOINT = ConfigManager.sharedInstance.darkSkyAPI + "/forecast/5136e36ec5498cfd8c583e3f580d3060"
```
Comment (6nop) [inline CPT331.iOS/CPT331.iOS/WeatherManager.swift]: Celebration

**Comment by 6nop** [comment]
**PR Approved**

0 Pull Requests remining

Celebration

**Pull request merged**